Docs upgrade warden (#550)
warden-protocol/wardenprotocol: docs/developer-docs/docs/operate-a-node/buenavista-testnet/upgrade/v0.4.0.md

:::warning
The upgrade procedure carries heightened risks of double-signing and being slashed. To handle these risks and securely address any issues that may arise during the upgrade, follow our tips listed in [Security and troubleshooting](#security--troubleshooting).
:::

## 1. Verify & back up

Before starting your validator and signing, verify your software version and the genesis file hash.

We also recommend to take a **full data snapshot** prior to the upgrade.

The method depends on your infrastructure, but generally you can achieve this by backing up the **wardend** directory. If you're [using Cosmovisor](#option-2-use-cosmovisor) to upgrade, it'll back up your data by default.

:::warning
(?) After stopping the `wardend` process, it's critically important to back up the `.wardend/data/priv_validator_state.json` file.
:::

**Note**: The **priv_validator_state.json**  file is updated every block as your validator participates in consensus rounds. In case the upgrade fails and you have to restart the previous chain, you'll need this file to avoid double signing. See [Handling node restarts](#handling-node-restarts).

## 2. Run Connect sidecar

> (?) Now you need to run the Skip:Connect binary on the same machine as `wardend`.

Reviewer: They can run their Skip:Connect before hand but they need to make sure they enable it from the configuration

Reviewer 2: Will it be helpful to back link to config files how to enable connect?

Author: @alijnmerchant21 @jlehtimaki  Where are these config files? I'm not sure I understand this addition. Does it sound correct?

Prior to the update, you should also run the Skip:Connect binary on the same machine as wardend. Make sure you enable it from the configuration.

Reviewer: Referring to this:
https://docs.wardenprotocol.org/operate-a-node/operate-connect#3-configure-wardend

> Make sure you enable oracle from Wardend app.toml configuration file.

Follow this guide: [Operate Connect](/operate-a-node/operate-connect).

## 3. Upgrade

### Option 1: Upgrade manually

1. (?) Download the [v0.4.0 wardend binary](https://github.com/warden-protocol/wardenprotocol/releases/tag/v0.4.0) or build it from the source code. See the installation instructions in the [Join Buenavista guide](/operate-a-node/buenavista-testnet/join-buenavista#1-install).

> 2. Run Warden v4.0.0 until the upgrade height. The node will panic:

Reviewer: No, this will result an error, they need to upgrade the binary `after` the upgrade height

Author: @jlehtimaki Could you please double check the section about manual upgrade, how does it look now? Is the error message correct?

Reviewer: looks good

> 2. Run Warden v4.0.0 until the upgrade height. The node will panic:

Reviewer: ```suggestion
2. Run Warden v0.3.0 until the upgrade height. The node will panic:
```

    (?)
    
    ```shell
    ERR UPGRADE "v0.4.0" NEEDED at height: 12255555: upgrade to v0.4.0 and applying upgrade "v0.4.0" at     height: 12255555
    ```

> 3. Stop the node and switch the binary to **Warden v4.0.1**. Restart the node:  

Reviewer: ```suggestion
3. Stop the node and switch the binary to **Warden v0.4.1**. Restart the node:  
```

    ```
    wardend start
    ```

> 4. The chain will continue producing blocks when validators with a total sum voting power > 2/3 complete their node upgrades. It may take from several minutes to a few hours.

Reviewer: ```suggestion
4. The chain will continue producing blocks when validators with a total sum voting power > 2/3 complete their node upgrades and come back online.
```

### Option 2: Use Cosmovisor

To upgrade using [Cosmovisor](https://pkg.go.dev/cosmossdk.io/tools/cosmovisor), take these steps:

> 1. Build Warden v0.4.0. and move the `wardend` v0.4.0 binary to `$WARDEN_HOME/cosmovisor/upgrades/v0.4.0/bin`:

Reviewer: Let's say that we want them to download the binary and not Build it

    
    ```shell
    mkdir -p  $WARDEN_HOME/cosmovisor/upgrades/v0.4.0/bin
    cp $(which wardend) $WARDEN_HOME/cosmovisor/upgrades/v0.4.0/bin
    ```

>     **Note**: Alternatively, enable auto-download in Cosmovisor, and it'll automatically download the binary.

Reviewer: Should we move this section to top?

Reviewer 2: > Should we move this section to top?

You mean using cosmovisor before manual update? Yes - I feel so too. Most people use cosmovisor to upgrade their binaries.

Reviewer: Yes that and before manual cosmovisor binary update

>     **Note**: Alternatively, enable auto-download in Cosmovisor, and it'll automatically download the binary.

Reviewer 2: Should we remove this maybe?

`**Note**: Alternatively, enable auto-download in Cosmovisor, and it'll automatically download the binary.`

Maybe its ok on testnets, but for mainnets it is not considered a secure practice to automatically download binaries..

Reviewer: No, I'm really arguing against that perspective, we ran 1 year without any issues the automatic update, works flawlessly when the binaries are listed in gov proposal.

2. Start the node:

    ```shell
    cosmovisor run start --x-crisis-skip-assert-invariants --home $DAEMON_HOME
    ```
    
    **Note**: We recommend skipping the invariant checks to reduce the upgrade time. Future releases of the Cosmos SDK will include additional improvements to the crisis module.

3. After reaching the upgrade block height, Warden will panic and stop.

> 4. The chain will continue producing blocks when validators with a total sum voting power > 2/3 complete their node upgrades. It may take from several minutes to a few hours.

Reviewer: ```suggestion
4. The chain will continue producing blocks when validators with a total sum voting power > 2/3 complete their node upgrades and come back online.
```

## Security & troubleshooting

:::warning
The upgrade procedure carries heightened risks of double-signing and being slashed. To mitigate these risks and securely address any issues that may arise during the upgrade, follow the rules listed below.
:::

### Handling mistakes

If you discover you've made a mistake, it's not safe repeat the upgrade procedure during the network startup. Instead, wait for the network to start and then correct the mistake.

### Handling node restarts

You node may restart during the upgrade due to OOM being killed, hardware issues, and other reasons. In this case, it's not safe to proceed with the upgrade. Instead, follow these steps:

1. Recover your node from a [backup](#1-verify--back-up).  

>     When applying a snapshot from the backup, keep your `./data/priv_validator_state.json` file. This will help you to avoid being slashed due to double signing.

Reviewer: We should say that they need to be careful also with their private node keys, those should be also backed up.
